### PR TITLE
Slice array with smaller mask

### PR DESCRIPTION
**Pull request by jakirkham**
Allows a Dask Array to be sliced with a boolean Dask Array that is smaller than the array it is slicing. This constitutes a very small, but useful step towards resolving issue ( https://github.com/dask/dask/issues/2651 ).

### CODE REVIEW

File: dask/array/core.py [hunk 1]
```diff
@@ -3213,12 +3213,27 @@ def from_npy_stack(dirname, mmap_mode='r'):
...
-    y = elemwise(getitem, x, index, dtype=x.dtype)
-
     name = 'getitem-' + tokenize(x, index)
 
-    dsk = {(name, i): k
-           for i, k in enumerate(core.flatten(y._keys()))}
-    chunks = ((np.nan,) * y.npartitions,)
+    if index.ndim < x.ndim:
+        x = x.reshape((-1,) + x.shape[index.ndim:])
+        index = index.flatten()
```
Comment (jakirkham) [inline dask/array/core.py]: This won't work if either array has an unknown length. Not sure of a good, simple way to fix this.

Comment (mrocklin) [reply dask/array/core.py]: This is, I think, something that we want to support.  the use case that I had in mind personally was for machine learning workloads, where the arrays almost always come from dataframes with unknown row-chunk sizes.  

I believe that by default atop will allow nan-chunked arrays to interact with each other assuming that they have equivalent chunk sizes

Comment (jakirkham) [reply dask/array/core.py]: The problem is reshaping doesn't seem to be supported on arrays with unknown dimension lengths. Is that something that could be supported? If not, I don't know of a simple fix.

**Comment by mrocklin** [comment]
Probably not easily?  I'm inclined just to wait until the full fix.  I
don't think this is critical near-term.

**Comment by jakirkham** [comment]
However the full fix to that issue is quite a bit of work (as already noted). The last attempt at solving it (before your more recent one) has been untouched for 3 months. This suggest to me that we should break this problem up into smaller more achievable chunks to make it easier for more people to push it forward. Do you agree?

I would think one problem is the one this PR tries to tackle. Namely allowing the mask to be smaller than the array it is applied to. There may be other ways to tackle it that are worth considering. Though I think solving a problem of this scope is a good first step.

Once we solve this problem, it should be possible to build off such a solution and consider where that mask is positioned. As a next step, we could consider allowing multiple different masks for some axes of the array being sliced.

It's my opinion that breaking this problem up will allow us to tackle it in a reasonable period of time instead of letting it idle for months. Plus it will be easier for interested contributors to pick up a piece of this work and complete it.

**Comment by mrocklin** [comment]
> This suggest to me that we should break this problem up into smaller more achievable chunks to make it easier for more people to push it forward. Do you agree?

In the general principle, yes, but this isn't the way in which I would break it up.  Instead I would make a change to `top` and `atop` to support literals.  I think that this will allow us to resolve this issue (and possibly several more) much more easily.

**Comment by mrocklin** [comment]
Should this be closed in favor of https://github.com/dask/dask/pull/2658 ?

**Comment by jakirkham** [comment]
Thanks for working on PR ( https://github.com/dask/dask/pull/2658 ), @mrocklin. That will be very useful.

In that PR bool arrays that are 1-D or the same dimensionality as the data were ok, but I don't think it was permissible to have a single bool array inbetween those two ranks (unless I'm missing something). The objective of this PR was to handle this inbetween case. That said, most of the cases I would want to handle are already well addressed by your PR.

Though if solving this case is useful to others, I could rework it in the context of your PR pretty easily as you have already done the hard work. Just thinking about this briefly would do a reshape of the data and mask to get the mask to actually be 1-D before applying it. That said, this is easy for anyone to do themselves too. So would be happy either reworking this or closing it. Please let me know your thoughts.

**Comment by jakirkham** [comment]
On second thought, I think this is just better off closed.

**Pull request closed without merging**